### ISSUE TITLE
multi-pozycje grid

### ISSUE THREAD

**Issue by Holender26** [Holender26/Scalp-bot#2]
cat > grid_multi_paper.py <<'PY'
import json, time, os, csv, math
from datetime import datetime, timezone
import requests
import argparse

BINANCE_PRICE_URL = "/api/v3/ticker/price"

def now_utc():
    return datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

def load_config(path):
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)

def to_binance_symbol(symbol: str) -> str:
    return symbol.replace("/", "").upper().strip()

def get_price(symbol_binance: str, base_url: str) -> float:
    url = base_url.rstrip("/") + BINANCE_PRICE_URL
    r = requests.get(url, params={"symbol": symbol_binance}, timeout=10)
    r.raise_for_status()
    return float(r.json()["price"])

def ensure_csv(path: str, header: list[str]):
    if os.path.exists(path):
        return
    with open(path, "w", newline="", encoding="utf-8") as f:
        csv.writer(f).writerow(header)

def append_csv(path: str, row: list):
    with open(path, "a", newline="", encoding="utf-8") as f:
        csv.writer(f).writerow(row)

def pick_active_strategy(strategies: dict) -> str:
    for name, cfg in strategies.items():
        if isinstance(cfg, dict) and cfg.get("enabled") is True:
            return name
    return "none"

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--config", required=True)
    args = ap.parse_args()

    cfg = load_config(args.config)

    mode = cfg.get("mode", "paper")
    if mode != "paper":
        print("This script is PAPER only. Set mode=paper in config.")
        return

    ex = cfg.get("exchange", {})
    base_url = ex.get("base_url", "https://api.binance.com")

    market = cfg.get("market", {})
    symbol = market.get("symbol", "BTC/USDT")
    symbol_b = to_binance_symbol(symbol)

    risk = cfg.get("risk", {})
    max_daily_loss_pct = float(risk.get("max_daily_loss_pct", 5.0))
    cooldown_after_loss = int(risk.get("cooldown_seconds_after_loss", 120))
    max_open_positions = int(risk.get("max_open_positions", 3))

    exec_cfg = cfg.get("execution", {})
    slippage_bps = float(exec_cfg.get("slippage_bps", 0))
    fee_bps = float(exec_cfg.get("fee_bps", 0))
    total_cost_bps = slippage_bps + fee_bps

    strategies = cfg.get("strategies", {})
    active = pick_active_strategy(strategies)
    if active != "grid":
        print(f"Active strategy is '{active}'. Enable strategies.grid.enabled=true in config.")
        return

    grid = strategies.get("grid", {})
    spacing_pct = float(grid.get("grid_spacing_pct", 0.6)) / 100.0
    levels = int(grid.get("levels", 14))
    take_profit_pct = float(grid.get("take_profit_pct", 0.4)) / 100.0
    global_stop_pct = float(grid.get("global_stop_pct", 6.0)) / 100.0

    trend_filter = grid.get("trend_filter", {}) or {}
    tf_enabled = bool(trend_filter.get("enabled", False))

    logging = cfg.get("logging", {})
    trades_csv = logging.get("trades_csv", "trades_12345.csv")
    events_csv = os.path.splitext(trades_csv)[0] + "_events.csv"

    ensure_csv(trades_csv, ["ts_utc","symbol","pos_id","side","qty","entry","exit","pnl_usdt","reason","level"])
    ensure_csv(events_csv, ["ts_utc","symbol","price","note"])

    # PAPER equity (prosto). Jak chcesz, podepnę pod config.
    equity_start = 1000.0
    equity = equity_start

    day_key = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    daily_pnl = 0.0
    cooldown_until = 0

    # Multi-pozycje
    positions = []  # list of dict: id, entry, qty, tp, stop, level
    next_pos_id = 1

    # Ustawienia grid
    px0 = get_price(symbol_b, base_url)
    ref = px0
    step = ref * spacing_pct

    # Poziomy BUY poniżej ref (najbliższy jako pierwszy)
    buy_levels = [ref - i * step for i in range(1, levels + 1)]
    buy_levels.sort(reverse=True)  # np. 0.6% poniżej, potem 1.2%, itd.

    append_csv(events_csv, [now_utc(), symbol_b, px0, f"START ref={ref:.2f} step={step:.2f} levels={levels} max_pos={max_open_positions} costs_bps={total_cost_bps}"])
    print(f"[{now_utc()}] START PAPER GRID-MULTI id={cfg.get('id')} symbol={symbol}({symbol_b})")
    print(f"[{now_utc()}] ref={ref:.2f} step={step:.2f} levels={levels} max_open_positions={max_open_positions}")
    print(f"[{now_utc()}] tp={take_profit_pct*100:.3f}% stop={global_stop_pct*100:.2f}% costs_bps={total_cost_bps} equity_start={equity_start}")

    tick = 0

    while True:
        try:
            # reset dzienny
            day_now = datetime.now(timezone.utc).strftime("%Y-%m-%d")
            if day_now != day_key:
                day_key = day_now
                daily_pnl = 0.0
                cooldown_until = 0
                append_csv(events_csv, [now_utc(), symbol_b, "", "daily_reset"])

            px = get_price(symbol_b, base_url)
            tick += 1

            if tick % 2 == 0:
                append_csv(events_csv, [now_utc(), symbol_b, px, f"feed pos_open={len(positions)}"])

            # limit dzienny straty
            if daily_pnl <= -abs(equity_start) * (max_daily_loss_pct / 100.0):
                append_csv(events_csv, [now_utc(), symbol_b, px, f"HALT daily_loss_limit pnl={daily_pnl:.2f}"])
                print(f"[{now_utc()}] HALT: daily loss limit hit. pnl={daily_pnl:.2f}")
                time.sleep(5)
                continue

            # cooldown po stracie
            if int(time.time()) < cooldown_until:
                time.sleep(2)
                continue

            # trend filter (bardzo prosta heurystyka)
            if tf_enabled:
                if abs(px - ref) > 2.5 * step:
                    append_csv(events_csv, [now_utc(), symbol_b, px, "trend_filter_pause"])
                    time.sleep(2)
                    continue

            # 1) Zarządzanie otwartymi pozycjami (TP/STOP)
            if positions:
                still_open = []
                for p in positions:
                    entry = p["entry"]
                    qty = p["qty"]
                    tp = p["tp"]
                    stop = p["stop"]

                    # STOP
                    if px <= stop:
                        exit_px = px * (1 - total_cost_bps / 10000.0)
                        pnl = (exit_px - entry) * qty
                        equity += pnl
                        daily_pnl += pnl
                        append_csv(trades_csv, [now_utc(), symbol_b, p["id"], "LONG", round(qty, 8), round(entry, 2), round(exit_px, 2), round(pnl, 4), "STOP", p["level"]])
                        append_csv(events_csv, [now_utc(), symbol_b, px, f"STOP pos_id={p['id']} pnl={pnl:.2f} equity={equity:.2f}"])
                        if pnl < 0:
                            cooldown_until = int(time.time()) + cooldown_after_loss
                        continue

                    # TP
                    if px >= tp:
                        exit_px = px * (1 - total_cost_bps / 10000.0)
                        pnl = (exit_px - entry) * qty
                        equity += pnl
                        daily_pnl += pnl
                        append_csv(trades_csv, [now_utc(), symbol_b, p["id"], "LONG", round(qty, 8), round(entry, 2), round(exit_px, 2), round(pnl, 4), "TP", p["level"]])
                        append_csv(events_csv, [now_utc(), symbol_b, px, f"TP pos_id={p['id']} pnl={pnl:.2f} equity={equity:.2f}"])
                        continue

                    still_open.append(p)

                positions = still_open

            # 2) Wejścia GRID (multi)
            # Otwieramy nowe pozycje, dopóki:
            # - są wolne sloty
            # - cena spada do kolejnych poziomów
            # - są jeszcze poziomy do wykorzystania
            while buy_levels and len(positions) < max_open_positions and px <= buy_levels[0]:
                level_price = buy_levels.pop(0)  # wykorzystujemy najbliższy poziom

                # notional: 0.5% equity na pozycję (prosto). Minimum 10 USDT.
                notional = max(10.0, equity * 0.005)
                qty = notional / px

                entry_px = px * (1 + total_cost_bps / 10000.0)  # koszt wejścia
                tp = entry_px * (1 + take_profit_pct)
                stop = entry_px * (1 - global_stop_pct)

                p = {
                    "id": next_pos_id,
                    "entry": entry_px,
                    "qty": qty,
                    "tp": tp,
                    "stop": stop,
                    "level": round(level_price, 2),
                }
                next_pos_id += 1
                positions.append(p)

                append_csv(trades_csv, [now_utc(), symbol_b, p["id"], "LONG", round(qty, 8), round(entry_px, 2), "", 0.0, "ENTER", p["level"]])
                append_csv(events_csv, [now_utc(), symbol_b, px, f"ENTER pos_id={p['id']} entry={entry_px:.2f} tp={tp:.2f} stop={stop:.2f} level={p['level']} open={len(positions)}"])
                print(f"[{now_utc()}] ENTER pos_id={p['id']} qty={qty:.6f} entry={entry_px:.2f} tp={tp:.2f} stop={stop:.2f} level={p['level']} open={len(positions)}")

            # 3) Recenter siatki, jeśli wybije daleko (żeby nie zostać bez poziomów)
            # Jeśli cena poszła wysoko i nie ma już poziomów, ustawiamy nowy ref i nowe poziomy.
            if not buy_levels and len(positions) == 0:
                ref = px
                step = ref * spacing_pct
                buy_levels = [ref - i * step for i in range(1, levels + 1)]
                buy_levels.sort(reverse=True)
                append_csv(events_csv, [now_utc(), symbol_b, px, f"RECENTER ref={ref:.2f} step={step:.2f}"])

            time.sleep(3)

        except KeyboardInterrupt:
            print("\nStopped (CTRL+C).")
            break
        except Exception as e:
            append_csv(events_csv, [now_utc(), symbol_b, "", f"ERROR {e}"])
            time.sleep(3)

if __name__ == "__main__":
    main()
PY
